src/meta/windows_utils.rs

```diff
@@ -0,0 +1,326 @@
+use std::io;
+use std::path::PathBuf;
+use std::ffi::{OsStr, OsString};
+use std::ptr::null_mut;
+use std::os::windows::ffi::{OsStringExt, OsStrExt};
+
+use winapi::um::winnt;
+use winapi::um::accctrl::TRUSTEE_W;
+use winapi::shared::winerror;
+use winapi::ctypes::c_void;
+
+use super::{Owner, Permissions};
+
+const BUF_SIZE: u32 = 256;
+
+
+pub fn get_file_data(path: &PathBuf) -> Result<(Owner, Permissions), io::Error> {
+    // Overall design:
+    // This function allocates some data with GetNamedSecurityInfoW,
+    // manipulates it only through WinAPI calls (treating the pointers as
+    // opaque) and then frees it at the end with LocalFree.
+    //
+    // For memory safety, the critical things are:
+    // - No pointer is valid before the return value of GetNamedSecurityInfoW
+    //   is checked
+    // - LocalFree must be called before returning
+    // - No pointer is valid after the call to LocalFree
+
+    let windows_path = buf_from_os(path.as_os_str());
+
+    // These pointers will be populated by GetNamedSecurityInfoW
+    // sd_ptr points at a new buffer that must be freed
+    // The others point at (opaque) things inside that buffer
+    let mut owner_sid_ptr = null_mut();
+    let mut group_sid_ptr = null_mut();
+    let mut dacl_ptr = null_mut();
+    let mut sd_ptr = null_mut();
+
+    // Assumptions:
+    // - windows_path is a null-terminated WTF-16-encoded string
+    // - The return value is checked against ERROR_SUCCESS before pointers are used
+    // - All pointers are opaque and should only be used with WinAPI calls
+    // - Pointers are only valid if their corresponding X_SECURITY_INFORMATION
+    //   flags are set
+    // - sd_ptr must be freed with LocalFree
+    let error_code = unsafe {
+        winapi::um::aclapi::GetNamedSecurityInfoW(
+            windows_path.as_ptr(),
+            winapi::um::accctrl::SE_FILE_OBJECT,
+            winnt::OWNER_SECURITY_INFORMATION
+                | winnt::GROUP_SECURITY_INFORMATION
+                | winnt::DACL_SECURITY_INFORMATION,
+            &mut owner_sid_ptr,
+            &mut group_sid_ptr,
+            &mut dacl_ptr,
+            null_mut(),
+            &mut sd_ptr
+        )
+    };
+
+    if error_code != winerror::ERROR_SUCCESS {
+        return Err(std::io::Error::from_raw_os_error(error_code as i32))
+    }
+
+    // Assumptions:
+    // - owner_sid_ptr is valid
+    // - group_sid_ptr is valid
+    // (both OK because GetNamedSecurityInfoW returned success)
+    let (owner_name, owner_domain) = unsafe { lookup_account_sid(owner_sid_ptr) }?;
```

Comment:
Propagating errors in this function will cause `sd_ptr` to be leaked. Safest option is to wrap `sd_ptr` in a struct that calls `LocalFree` in its drop method. You won't have to worry about making sure it's cleaned up in every path.